entoas/edges.go

```diff
@@ -0,0 +1,117 @@
+package entoas
+
+import (
+	"fmt"
+
+	"entgo.io/contrib/entoas/serialization"
+	"entgo.io/ent/entc/gen"
+)
+
+const _maxDepth = 25
+
+type (
+	// Edge wraps a gen.Edge and denotes an edge to be returned in an operation response. It recursively defines
+	// edges to load on the gen.Type the wrapped edge is pointing at.
+	Edge struct {
+		*gen.Edge
+		Edges Edges
+	}
+	Edges []*Edge
+	// walk is a node sequence in the schema graph. Used to keep track when creating an Edge.
+	walk []string
+)
+
+// EdgeTree returns the Edges to include on a type for the given serialization groups.
+func EdgeTree(n *gen.Type, gs serialization.Groups) (Edges, error) { return edgeTree(n, walk{}, gs) }
+
+// Flatten returns a list of all gen.Types present in the tree.
+func (es Edges) Flatten() []*gen.Edge {
+	em := make(map[string]*gen.Edge)
+	types(em, es)
+	var r []*gen.Edge
+	for _, t := range em {
+		r = append(r, t)
+	}
+	return r
+}
+
+// types recursively adds all gen.Types present on the given Edges to the given map.
+func types(dest map[string]*gen.Edge, es Edges) {
+	for _, e := range es {
+		dest[e.Type.Name] = e.Edge
+		types(dest, e.Edges)
+	}
+}
+
+// edgeTree recursively collects the edges to load on this type for the requested groups.
+func edgeTree(n *gen.Type, w walk, gs serialization.Groups) (Edges, error) {
+	// If we have reached maxDepth there most possibly is an unwanted circular reference.
+	if w.reachedMaxDepth() {
+		return nil, fmt.Errorf("max depth of %d reached: ", _maxDepth)
+	}
+	// Iterate over the edges of the given type.
+	// If the type has an edge we need to eager load, do so.
+	// Recursively go down the current types edges and, if requested, eager load those too.
+	var es Edges
+	for _, e := range n.Edges {
+		a, err := GetEdgeAnnotation(e)
+		if err != nil {
+			return nil, err
+		}
+		if a.MaxDepth == 0 {
+			a.MaxDepth = 1
+		}
+		// If the edge has at least one of the groups requested, load the edge.
+		if a.Groups.Match(gs) {
+			// Add the current step to our walk, since we will add this edge.
+			w.push(n.Name + "." + e.Name)
+			// If we have reached the max depth on this field for the given type stop the recursion. Backtrack!
+			if w.cycleDepth() > a.MaxDepth {
```

Comment:
> // If we have reached the max depth on this field for the **given type** stop the recursion. 

How does it work if you push `n.Name + "." + e.Name` (e.g. `User.pets`) and not `n.Name` (e.g. `User`)?

Reply:
I am tracking the cycle alongside an Graph. For example if I have `User`->hasMany(firends)->`User` and `User`->hasMany(enemies)->`User` I have to track `User.friends` and `User.enemies`. If I only do `User` I don't know which edge led me there.

Reply:
Well, starting from an entity `depth` is the amount of edges we walked along. Can be called something else like `length` or what. But if you imagine the object-graph of a schema-graph (a state on the DB) `depth` makes sense.